bridge/server/user/oauth-routes.ts

```diff
@@ -0,0 +1,162 @@
+const codeVerifiers: { [state: string]: { codeVerifier: string; nonce: string; expiresAt: number } } = {};
+const stateExpireMilliSeconds = 60 * 60_000; // expires in 60 minutes
+
+/**
+ * Build the root path. The exact path depends on the deployment & PREFIX_PATH value
+ *
+ * If PREFIX_PATH is defined, root location is set to <PREFIX_PATH>/bridge. Otherwise, root is set to / .
+ *
+ * Redirection to / will be either handled by Nginx (ex:- generic keptn deployment) OR the Express layer (ex:- local bridge development).
+ */
+function getRootLocation(): string {
+  if (prefixPath !== undefined) {
+    return `${prefixPath}/bridge`;
+  }
+  return '/';
+}
+
+function oauthRouter(client: BaseClient, redirectUri: string, reduceRefreshDateSeconds: number): Router {
+  const router = Router();
+  const additionalScopes = process.env.OAUTH_SCOPE ? ` ${process.env.OAUTH_SCOPE.trim()}` : '';
+  const scope = `openid${additionalScopes}`;
+
+  /**
+   * Router level middleware for login
+   */
+  router.get('/login', async (req: Request, res: Response) => {
+    const codeVerifier = generators.codeVerifier();
+    const codeChallenge = generators.codeChallenge(codeVerifier);
+    const nonce = generators.nonce();
+    const state = generators.state();
+    codeVerifiers[state] = { codeVerifier, nonce, expiresAt: new Date().getTime() + stateExpireMilliSeconds };
+
+    const authorizationUrl = client.authorizationUrl({
+      scope,
+      state,
+      nonce,
+      code_challenge: codeChallenge,
+      code_challenge_method: 'S256',
+    });
+
+    res.redirect(authorizationUrl);
+    return res;
+  });
+
+  /**
+   * Router level middleware for redirect handling
+   */
+  router.get('/oauth/redirect', async (req: Request, res: Response) => {
+    const params = client.callbackParams(req);
+
+    if (!params.code || !params.state) {
+      return res.redirect(getRootLocation());
+    }
+
+    const verifiers = codeVerifiers[params.state];
+    if (verifiers) {
+      delete codeVerifiers[params.state];
+      params.state = undefined;
+    } else {
+      return res.render('error', {
+        title: 'Permission denied',
+        message: 'Forbidden',
+      });
+    }
+
+    try {
+      const tokenSet = await client.callback(redirectUri, params, {
+        code_verifier: verifiers.codeVerifier,
+        nonce: verifiers.nonce,
+        scope,
+      });
+      reduceRefreshDateBy(tokenSet, reduceRefreshDateSeconds);
+      await authenticateSession(req, tokenSet);
+      res.redirect(getRootLocation());
+    } catch (error) {
+      const err = error as errors.OPError | errors.RPError;
+      console.log(`Error while handling the redirect. Cause : ${err.message}`);
+
+      if (err.response?.statusCode === 403) {
+        const response = {
+          title: 'Permission denied',
+          message: '',
+        };
+        response.message =
+          (err.response.body as Record<string, string>).message ?? 'User is not allowed access the instance.';
```

Comment:
Just a "to" missing here. easy small change: 
```suggestion
          (err.response.body as Record<string, string>).message ?? 'User is not allowed to access the instance.';
```